src/op_handlers/far_call.rs

```diff
@@ -1,14 +1,21 @@
-use zkevm_opcode_defs::FarCallOpcode;
-
-use crate::state::VMState;
-
-pub fn far_call(vm: &mut VMState, opcode: &FarCallOpcode) {
-    match opcode {
-        FarCallOpcode::Normal => {
-            let program_code = vm.current_frame().code_page.clone();
-            let stipend = vm.current_frame().gas_left;
-            vm.push_far_call_frame(program_code, stipend.0 / 32);
-        }
-        _ => todo!(),
-    }
+use u256::U256;
+use zkevm_opcode_defs::ethereum_types::Address;
+
+use crate::{state::VMState, Opcode};
+
+fn u256_to_address(integer: &U256) -> Address {
+    let mut buffer = [0u8; 32];
+    integer.to_big_endian(&mut buffer);
+
+    Address::from_slice(&buffer[12..32])
+}
+
+pub fn _far_call_normal(vm: &mut VMState, opcode: &Opcode) {
```

Comment:
Restore this to the far_call function and match on the opcode to call this one.
lib.rs should still call far_call.

Reply:
No, what I mean is to restore the far_call function like it was previously.
Then, call far_call_normal inside that one.
AFAIK, each far_call variant has the same setup, give or take, so I see value in having everything in one place, at least for now.

Reply:
I see, I've reverted this changes, but there are some changes needed tho, like passing the complete `Opcode` as an argument instead of only the type of `FarCallOpcode`, to access fields such as `src0_index`.

```rs
pub fn far_call(vm: &mut VMState, opcode: &Opcode, far_call_variant: FarCallOpcode) {
    match far_call_variant {
        FarCallOpcode::Normal => {
            let program_code = vm.current_frame().code_page.clone();
            let stipend = vm.current_frame().gas_left;
            let address_mask: U256 = U256::MAX >> (256 - 160);
            let address =
                u256_to_address(&((vm.get_register(opcode.src0_index).value) & address_mask));
            let caller = vm.current_context().contract_address;
            vm.push_far_call_frame(program_code, stipend.0 / 32, address, caller);
        }
        _ => todo!(),
    }
}
```